### ISSUE TITLE
Fix Spack Packages Without Checksummable Versions

### ISSUE THREAD

**Issue by citibeth** [LLNL/spack#3112]
@tgamblin @becker33 

@adamjstewart wrote:
> There are dozens of packages currently in Spack with no checksummed versions.

Then that is a serious problem and needs to be fixed.

Normally, Spack is "safe," and *all installs are checksummed*.  This is a key selling point for people advocating for Spack within their institutions: if site managers believe that Spack could introduce vulnerabilities to their system, they will not allow it.

Unfortunately... in spite of all the work that's point into checksums... in the current version of Spack it can all be defeated simply by adding a package with no checksummable version (and getting users to install it).  This booby-trap is a serious problem, and cuts against the message we want to project, that Spack is safe.

Until we have a better security system in place, our best defense against this problem is to require that all packages have at least one checksummable release.

Even *with* a better security system, it will still be advisable to have at least one checksummable release in every package --- because Spack won't be able to install the package without it, unless users turn off the security measures (which they shouldn't).

**Comment by citibeth** [comment]
TODO: Someone needs to identify packages that don't have at least one checksummable versions.  With that list, then we can decide what (more) to do about them.

**Comment by adamjstewart** [comment]
What is unsafe about git checkouts?

**Comment by citibeth** [comment]
The crux of Spack security is that the tarball you see when you download is the same as the tarball originally seen by the package author.  Checksums ensure this property.  It's a very strong property: it no longer matters who runs the website you downloaded things from, or whether `https://` was used, or whether they've been hacked, or whether you got it from a friend of a friend on a USB stick they found in the street.  As long as the tarball checksums properly, you know it's good.

If you do a `git` download from a named branch or tag, you have no such guarantee.  Now, you have to ask much more complex (and unanswerable) questions to know that you're getting the same tarball as the package author.  Has the git server been compromised?  Has the branch/tag moved since the package was created?  Even if there is no security breach, a git branch/tag moving could make it hard to repeat / resolve bugs.  You don't really know what version you installed.

I believe that `git` downloads specified by a Git hash (*not* a named branch or tag) avoid these problems.


Checksums ensure that the download you see is the same as the download seen by the package author who wrote the checksum into `package.py`.

**Comment by adamjstewart** [comment]
I understand your concerns, but for a lot of software, releases are not issued. The developers intend and recommend that users always checkout the latest master branch, because they fix bugs daily and don't want to have to debug something that they've already fixed. An example is https://github.com/idaholab/moose, which I tried packaging a while back. Although there was some random release once upon a time, the developers recommend installing the latest master branch.

By adding that random release, we now say that _Spack recommends_ installing something other than what the _developers recommend_. I don't like that.

@tgamblin I thought we trusted git checkouts as secure?

**Comment by citibeth** [comment]
> but for a lot of software, releases are not issued. The developers intend and recommend that users always checkout the latest master branch, because they fix bugs daily and don't want to have to debug something that they've already fixed.  An example is https://github.com/idaholab/moose, which I tried packaging a while back. Although there was some random release once upon a time, the developers recommend installing the latest master branch.

Yes, I work with such software as well.  Since it is impossible to checksum such packages, users should be aware when they are using them.  There's nothing wrong with a user deciding "I'm downloading from a git branch; but I know who these developers are, and I trust their code is not malicious, so I'm willing to download without checksums."  I do that too.

But `spack install` can install a lot of stuff automatically, without much awareness on the user's part.  The problem is when unchecksummed code is downloaded *without knowledge, understanding or consent* on the part of the user.  Understanding and consent is given when the user puts `version: [develop]` in their `packages.yaml` file.

Let's focus on: **What should Spack do if a package has no checksummable versions and the user has not given consent to install an unchecksummable version?**  In an ideal world, I believe you are right --- Spack should give an error and stop.  That would be a great PR.  But until we have that feature, we still need to stop Spack from automatically installing unchecksummable versions without user consent.  Currently the only way we have to do that is to add a release by git-hash.

> By adding that random release, we now say that Spack recommends installing something other than what the developers recommend. I don't like that.

I'd be happy to add a comment next to the "release" explaining that we don't really recommend it, and that users should give consent by adding to their `packages.yaml`.

> @tgamblin I thought we trusted git checkouts as secure?

Only if we checkout by git hash, not branch/tag.

**Comment by citibeth** [comment]
In a review of #2972 I found the following issues regarding uncheksummable packages:

No release
--------------
These packages do not have any checksummable releases.  Without choosing a release, there is no way to "safely" install them for users:
```
cleverleaf
hdf5-blosc
openspeedshop
the_platinum_searcher
```

Should have Git hash
-----------------------
These packages are downloading from release tags, which should be stable.  Git hashes need to be added to these downloads, to be equivalent to a tarball download + checksum (right now they're like a tarball download without checksum).
```
damselfly
mitos
py-yt
rust
tumxinator
```
BTW... what happens if a `tag=` and `hash=` are added to a Git download method simultaneously?  I believe that Spack should verify that the tag matches the hash, and download only if they match.  That would be a nice way to keep things consistent.

Misc
-----

* The `mfem` package downloads from a `goo.gl` URL.  how strange...  Although this is not a security problem, it should be fixed to a real URL.

**Comment by becker33** [comment]
@psakievich do you agree that this will be fixed by #44319 ? Can we mark that in the description for #44319 if so?

**Comment by psakievich** [comment]
I think it needs to be fixed in the scope of #44319 but I was seeing it as a separate PR.  One way or another this exact issue needs to be fixed as part of our binary provenance effort. 

It will solve the later part of the git tag matching the commit hash, but I think the broader aspect of tarballs and mirrors is hit on in this issue as well.
